### PR TITLE
Add test for tcms.rpc.api.testscase #1623

### LINKED ISSUE #1623: Automated tests for missing coverage in tcms.rpc.api.testcase
Kiwi TCMS needs additional automated test(s) as mentioned in subject.

**Definition of Done**:
- full coverage for the module or class mentioned in the title. This means both positive/negative scenarios where this makes sense or scenarios which traverse different branches of the code where that makes sense.
- test class/test method names are descriptive and reflect the behavior of the subject under test.  A good scheme to follow is `Test<ClassUnderTest>.test_[who][what][expected_results]`
- CI pass
- All code review passed & PR merged

Hints:
- Coverage report is at https://codecov.io/gh/kiwitcms/Kiwi/tree/master/tcms
- Devel getting started docs: https://kiwitcms.readthedocs.io/en/latest/contribution.html#testing
- Django testing documentation: https://docs.djangoproject.com/en/3.0/topics/testing/

*This issue is part of Kiwi TCMS open source bounty program. For more information see the link(s) in [bounty-program milestone](https://github.com/kiwitcms/Kiwi/milestone/4)*

### PR DESCRIPTION
**Pull request by pandafy**
Fixes #1623

### CODE REVIEW

**Review by atodorov — CHANGES_REQUESTED** [review]
`Fixes #1623` goes onto the first line of the commit log, after the actual message

File: tcms/rpc/tests/test_testcase.py [hunk 1]
```diff
@@ -576,6 +749,10 @@ def verify_api_with_permission(self):
         self.assertEqual(result[2]['case_id'], self.case_3.pk)
         self.assertEqual(result[2]['sortkey'], 25)
 
+        # Test with query none
+        result = self.rpc_client.TestCase.sortkeys()
+        self.assertDictEqual(result, {})
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:754]: This is causing failures b/c your assertion is wrong. When the filter dict is empty all objects in the corresponding DB table will be returned. So that is all sortkeys for all test cases from all test plans that currently exist.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/rpc/tests/test_testcase.py (hunk 1 above)
Comment (pandafy) [reply tcms/rpc/tests/test_testcase.py:754]: Got it! I assumed it is `return {}`instead of `query = None`. :sweat_smile:

**Comment by pandafy** [comment]
> `Fixes #1623` goes onto the first line of the commit log, after the actual message

@atodorov can we add instructions for writing commit messages in the contribution guidelines. This will help new contributors on how to structure their commit message. I can open up an issue for this if you think it is worth doing. 

Following are some commit message guidelines which I found out to be good:
- [Django's committing guideline](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines) 
- [OpenWISP's commit message guideline](https://openwisp.io/docs/developer/contributing.html#how-to-commit-your-changes-properly)

File: tcms/rpc/tests/test_testcase.py [hunk 2]
```diff
@@ -610,3 +700,49 @@ def verify_api_with_permission(self):
...
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.case = TestCaseFactory()
+        add_attachment(self.case.pk,
+                       'testcases.TestCase',
+                       self.tester,
+                       'test.txt',
+                       'test')
```
Comment (pandafy) [inline tcms/rpc/tests/test_testcase.py:738]: @atodorov how should I add an attachment here. The attachments does not shows up when the code below. What am I missing here? 

I looked into the implementation of `Testcase.add_attachment` and it used this method  `tcms.rpc.utils.add_attachment`.  I am calling this method as it would have been called in `Testcase.add_attachment`. 
Using an API method here does not makes any sense as you shared in your previous review.

**Comment by atodorov** [comment]
> > `Fixes #1623` goes onto the first line of the commit log, after the actual message
> 
> @atodorov can we add instructions for writing commit messages in the contribution guidelines. This will help new contributors on how to structure their commit message. I can open up an issue for this if you think it is worth doing.
> 
> Following are some commit message guidelines which I found out to be good:
> 
>     * [Django's committing guideline](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines)
> 
>     * [OpenWISP's commit message guideline](https://openwisp.io/docs/developer/contributing.html#how-to-commit-your-changes-properly)

We usually refer to this guideline - https://chris.beams.io/posts/git-commit/ - I haven't read the others but they are probably just as good. Feel free to open a pull request against the documentation, see `docs/source/contribution.rst`

File: tcms/rpc/tests/test_testcase.py (hunk 2 above)
Comment (atodorov) [reply tcms/rpc/tests/test_testcase.py:784]: Other than the fact that the last argument should be base64 encoded I don't see anything else that may be wrong. Maybe there is a silent failure here.  See how data is passed in  tests/test_testplan.py from the API method to the utils function.

File: tcms/rpc/tests/test_testcase.py [hunk 3]
```diff
@@ -3,26 +3,86 @@
...
+
+        self.default_cc = 'example@example.com'
+
+    def test_non_list_email_collection(self):
+        with self.assertRaisesRegex(TypeError, 'cc_list should be a list object.'):
+            _validate_cc_list((self.default_cc,))
+
+    def test_invalid_email(self):
+        with self.assertRaises(ValidationError):
+            _validate_cc_list(['@example.com'])
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:36]: This also needs a scenario where you have a list of addresses and all of them are valid.

File: tcms/rpc/tests/test_testcase.py [hunk 4]
```diff
@@ -3,26 +3,86 @@
...
+        self.assertEqual(len(self.testcase.emailing.get_cc_list()), 1)
+        self.assertEqual(self.testcase.emailing.get_cc_list(), [self.default_cc])
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestCase.add_notification_cc(self.testcase.pk, [self.default_cc])
+
+
+class TestGetNotificationCCPermission(APIPermissionsTestCase):
+    """ Tests the XML-RPC Testcase.get_notification_cc method """
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:59]: API method names are case sensitive so if you are going to document them then use the same case as in the method doc-string. Also makes searching for things easier.

File: tcms/rpc/tests/test_testcase.py [hunk 5]
```diff
@@ -3,26 +3,86 @@
...
+        cc_list = self.rpc_client.TestCase.get_notification_cc(self.testcase.pk)
+        self.assertListEqual(cc_list, [self.default_cc])
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestCase.get_notification_cc(self.testcase.pk)
+
+
+class TestRemoveNotificationCC(APITestCase):
+    """ Tests the XML-RPC testcase.remove_notification_cc method """
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:80]: API method names are case sensitive so if you are going to document them then use the same case as in the method doc-string. Also makes searching for things easier.

File: tcms/rpc/tests/test_testcase.py [hunk 6]
```diff
@@ -3,26 +3,86 @@
...
-    """ Tests the XML-RPC testcase.notication_remove_cc method """
+class TestValidateEmail(DjangoBaseTestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.default_cc = 'example@example.com'
+
+    def test_non_list_email_collection(self):
+        with self.assertRaisesRegex(TypeError, 'cc_list should be a list object.'):
+            _validate_cc_list((self.default_cc,))
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:32]: You can pass a hard-coded string value here, like you do below and will not need the setup method at all.

File: tcms/rpc/tests/test_testcase.py [hunk 7]
```diff
@@ -3,26 +3,86 @@
 
 from xmlrpc.client import Fault, ProtocolError
 
+from attachments.models import Attachment
 from django.contrib.auth.models import Permission
+from django.core.exceptions import ValidationError
+from django.test import TestCase as DjangoBaseTestCase
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:9]: `from django import test` and then use `test.TestCase` where needed.

File: tcms/rpc/tests/test_testcase.py [hunk 8]
```diff
@@ -3,26 +3,86 @@
...
 from tcms.testcases.models import Category, TestCase, TestCaseStatus
 from tcms.tests import remove_perm_from_user
 from tcms.tests.factories import (CategoryFactory, ComponentFactory,
                                   ProductFactory, TagFactory, TestCaseFactory,
                                   TestPlanFactory, UserFactory, VersionFactory)
 
 
-class TestNotificationRemoveCC(APITestCase):
-    """ Tests the XML-RPC testcase.notication_remove_cc method """
+class TestValidateEmail(DjangoBaseTestCase):
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:24]: Does this work if you inherit from `unittest.TestCase` instead ?

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/rpc/tests/test_testcase.py [hunk 9]
```diff
@@ -16,13 +21,66 @@
...
+    def verify_api_with_permission(self):
+        cc_list = self.rpc_client.TestCase.get_notification_cc(self.testcase.pk)
+        self.assertListEqual(cc_list, [self.default_cc])
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestCase.get_notification_cc(self.testcase.pk)
+
+
+class TestRemoveNotificationCC(APITestCase):
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:77]: Try converting to an APIPermissionsTestCase like you have done above which should cover the required permissions as well.

File: tcms/rpc/tests/test_testcase.py [hunk 10]
```diff
@@ -16,13 +21,66 @@
...
+    """ Tests the XML-RPC TestCase.remove_notification_cc method """
 
     def _fixture_setup(self):
         super()._fixture_setup()
 
-        self.default_cc = 'example@MrSenko.com'
+        self.default_cc = 'example@example.com'
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:83]: Missing scenario with invalid TC ID which asserts we get a DoesNotExist exception. See above.

File: tcms/rpc/tests/test_testcase.py [hunk 11]
```diff
@@ -430,6 +493,27 @@ def test_fails_when_mandatory_fields_not_specified(self):
...
+        super()._fixture_setup()
+
+        self.case_1 = TestCaseFactory()
+        self.case_2 = TestCaseFactory()
+        self.case_3 = TestCaseFactory()
+
+        self.query = {'pk__in': [self.case_1.pk, self.case_2.pk, self.case_3.pk]}
+
+    def verify_api_with_permission(self):
+        len_deleted, _ = self.rpc_client.TestCase.remove(self.query)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:509]: What's the second return value? IMO you can remove only 1 of these records and assert on the second value as well b/c API clients may be using that value.

File: tcms/rpc/tests/test_testcase.py [hunk 12]
```diff
@@ -522,7 +606,26 @@ def test_add_component_from_another_product_is_not_allowed(self):
...
 
-class TestCaseAddComment(APITestCase):
+class TestRemoveComponentPermission(APIPermissionsTestCase):
+    permission_label = 'testcases.delete_testcasecomponent'
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+        self.test_case = TestCaseFactory()
+        self.good_component = ComponentFactory(product=self.test_case.category.product)
+        self.bad_component = ComponentFactory()
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:616]: `bad_component` is unused. You can assign it to the same TC and then assert that it is the only one left standing.

File: tcms/rpc/tests/test_testcase.py [hunk 13]
```diff
@@ -522,7 +606,26 @@ def test_add_component_from_another_product_is_not_allowed(self):
...
+
+    def verify_api_with_permission(self):
+        self.rpc_client.TestCase.remove_component(self.test_case.pk,
+                                                  self.good_component.pk)
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestCase.remove_component(self.test_case.pk,
+                                                      self.good_component.pk)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:625]: Missing scenario with invalid TC ID which asserts we get a DoesNotExist exception. See above.

File: tcms/rpc/tests/test_testcase.py [hunk 14]
```diff
@@ -522,7 +606,26 @@ def test_add_component_from_another_product_is_not_allowed(self):
...
+
+
+class TestAddComment(APITestCase):
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:628]: This should inherit from APIPermissionsTestCase instead and follow the examples above.

File: tcms/rpc/tests/test_testcase.py [hunk 15]
```diff
@@ -541,6 +644,31 @@ def test_add_comment_with_pk_as_int(self):
...
+        self.rpc_client.TestCase.remove_comment(self.case.pk, self.comment_1.pk)
+        self.assertEqual(len(comments.get_comments(self.case)), 2)
+
+        # Remove all comments
+        self.rpc_client.TestCase.remove_comment(self.case.pk)
+        self.assertEqual(len(comments.get_comments(self.case)), 0)
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestCase.remove_comment(self.case.pk)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:669]: Missing scenario with invalid TC ID which asserts we get a DoesNotExist exception. See above.

File: tcms/rpc/tests/test_testcase.py [hunk 16]
```diff
@@ -569,14 +697,19 @@ def verify_api_with_permission(self):
         self.assertEqual(result[str(self.case_2.pk)], 15)
         self.assertEqual(result[str(self.case_3.pk)], 25)
 
+        # Test query None
+        result = self.rpc_client.TestCase.sortkeys(None)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:701]: You'd better call the method as `TestCase.sortkeys()` which makes sure to also use the specified default value in the method signature. Same applies for `TestCase.filter()` method.

File: tcms/rpc/tests/test_testcase.py [hunk 17]
```diff
@@ -603,3 +736,59 @@ def verify_api_with_permission(self):
...
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.case = TestCaseFactory()
+
+    def verify_api_with_permission(self):
+        self.rpc_client.TestCase.add_attachment(self.case.pk,
+                                                'test.txt',
+                                                'test')
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:752]: last parameter should be a base64 encoded string, see below. Otherwise we may be hitting unrelated errors.

File: tcms/rpc/tests/test_testcase.py [hunk 18]
```diff
@@ -603,3 +736,59 @@ def verify_api_with_permission(self):
...
+        attachments = Attachment.objects.attachments_for_object(self.case)
+        self.assertEqual(len(attachments), 1)
+        self.assertEqual(attachments[0].object_id, str(self.case.pk))
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestCase.add_attachment(self.case.pk,
+                                                    'test.txt',
+                                                    'test')
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:761]: last parameter should be a base64 encoded string, see below. Otherwise we may be hitting unrelated errors.

File: tcms/rpc/tests/test_testcase.py [hunk 19]
```diff
@@ -603,3 +736,59 @@ def verify_api_with_permission(self):
...
+        self.rpc_client.TestCase.add_attachment(self.case.pk, 'attachment.txt', 'a2l3aXRjbXM=')
+        attachments = self.rpc_client.TestCase.list_attachments(self.case.pk)
+        self.assertEqual(len(attachments), 1)
+
+    def test_list_attachments_with_wrong_case_id(self):
+        with self.assertRaisesRegex(Fault, 'TestCase matching query does not exist'):
+            self.rpc_client.TestCase.list_attachments(-1)
+
+
+class TestListAttachmentPermissions(APIPermissionsTestCase):
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:780]: You can merge TestListAttachments class with this one and validate both the permissions and the functionality in one place. For simpler methods that works nicely. 

If you don't want to change it it can stay like that as well.

Note the missing scenario for non existing TC id! In case you aren't changing this class then add it to TestListAttachments.

**Review by atodorov — CHANGES_REQUESTED** [review]
Test coverage is good except several corner cases and the other changes requested.

File: tcms/rpc/tests/test_testcase.py (hunk 11 above)
Comment (pandafy) [reply tcms/rpc/tests/test_testcase.py:509]: `TestCase.remove()` returns `TestCase.objects.filter(**query).delete()`. According to [Django's documentation](https://docs.djangoproject.com/en/3.1/topics/db/queries/#deleting-objects), `delete` method returns the number of objects deleted and a dictionary with the number of deletions per object type. 

I don't think the number of deletions per object will be helpful for our testing.

File: tcms/rpc/tests/test_testcase.py (hunk 19 above)
Comment (pandafy) [reply tcms/rpc/tests/test_testcase.py:780]: I took the reference from `tcms/rpc/tests/test_testplan.py` to implement tests for `list_attachments`. It seems like there is no easy way to add an attachment to a `TestCase` or a `TestPlan`. 

The sole reason for having two separate tests one for permissions and other for testing the functionality is the requirement of `attachments.add_attachment` permissions for adding an attachment which is not feasible in an instance of `APIPermissionsTestCase` class.

File: tcms/rpc/tests/test_testcase.py [hunk 20]
```diff
@@ -16,13 +21,66 @@
...
+        self.testcase = TestCaseFactory()
+
+    def verify_api_with_permission(self):
+        self.rpc_client.TestCase.add_notification_cc(self.testcase.pk, [self.default_cc])
+        self.assertEqual(len(self.testcase.emailing.get_cc_list()), 1)
+        self.assertEqual(self.testcase.emailing.get_cc_list(), [self.default_cc])
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestCase.add_notification_cc(self.testcase.pk, [self.default_cc])
```
[the remark this answers is not in the document]
Comment (pandafy) [reply tcms/rpc/tests/test_testcase.py:54]: I have added another class which decends from`APITestCase` to test the the corner case when `TestCase` id is invalid. Those tests were not working from descendants of `APIPermissionsTestCase`. 

Anther way to implement it would be to use `subTest` inside `verify_api_with_permission` function and test the corner case of TestCase id being invalid in one `subTest` and normal functionality with a valid `TestCase` id in another.

File: tcms/rpc/tests/test_testcase.py [hunk 21]
```diff
@@ -430,6 +508,27 @@ def test_fails_when_mandatory_fields_not_specified(self):
...
+        super()._fixture_setup()
+
+        self.case_1 = TestCaseFactory()
+        self.case_2 = TestCaseFactory()
+        self.case_3 = TestCaseFactory()
+
+        self.query = {'pk__in': [self.case_1.pk, self.case_2.pk, self.case_3.pk]}
+
+    def verify_api_with_permission(self):
+        len_deleted, _ = self.rpc_client.TestCase.remove(self.query)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:524]: A more appropriate name is `num_deleted` b/c it is a number after all.

File: tcms/rpc/tests/test_testcase.py [hunk 22]
```diff
@@ -522,16 +621,47 @@ def test_add_component_from_another_product_is_not_allowed(self):
...
+        self.test_case = TestCaseFactory()
+        self.good_component = ComponentFactory(product=self.test_case.category.product)
+        self.bad_component = ComponentFactory(product=self.test_case.category.product)
+        self.test_case.add_component(self.good_component)
+        self.test_case.add_component(self.bad_component)
+
+    def verify_api_with_permission(self):
+        self.rpc_client.TestCase.remove_component(self.test_case.pk,
+                                                  self.bad_component.pk)
+        result = TestCaseComponent.objects.filter(case=self.test_case)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:638]: `self.test_case.component` is a Django queryset (b/c there is a many-to-many relationship between these fields), which means it has methods .all(), .count() and .first().  So you can fetch the results after the deletion directly from the DB instead of going through the API layer.

File: tcms/rpc/tests/test_testcase.py [hunk 23]
```diff
@@ -64,6 +137,11 @@ def _fixture_setup(self):
                 plan=[self.plan])
             self.cases.append(test_case)
 
+    def test_filter_query_none(self):
+        result = self.rpc_client.TestCase.filter()
+        self.assertIsNotNone(result)
+        self.assertEqual(len(result), 10)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:143]: Convert the second assertion b/c it will break the moment there are more temporary objects into the database.  Check that it is not an empty list, that would be sufficient.

File: tcms/rpc/tests/test_testcase.py [hunk 24]
```diff
@@ -569,14 +741,19 @@ def verify_api_with_permission(self):
         self.assertEqual(result[str(self.case_2.pk)], 15)
         self.assertEqual(result[str(self.case_3.pk)], 25)
 
+        # Test query None
+        result = self.rpc_client.TestCase.sortkeys()
+        self.assertIsNotNone(result)
+        self.assertEqual(len(result), 3)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:747]: Same comment as above - will break the moment there are more objects into the database. Check if the list is not empty instead.

**Review by atodorov — CHANGES_REQUESTED** [review]
A few more changes requested.

File: tcms/rpc/tests/test_testcase.py [hunk 25]
```diff
@@ -430,6 +507,27 @@ def test_fails_when_mandatory_fields_not_specified(self):
...
+        super()._fixture_setup()
+
+        self.case_1 = TestCaseFactory()
+        self.case_2 = TestCaseFactory()
+        self.case_3 = TestCaseFactory()
+
+        self.query = {'pk__in': [self.case_1.pk, self.case_2.pk, self.case_3.pk]}
+
+    def verify_api_with_permission(self):
+        len_deleted, num_deleted = self.rpc_client.TestCase.remove(self.query)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:523]: `num_deleted` is the first return value, the second one was unused, right ?

File: tcms/rpc/tests/test_testcase.py [hunk 26]
```diff
@@ -522,16 +620,47 @@ def test_add_component_from_another_product_is_not_allowed(self):
...
+        self.bad_component = ComponentFactory(product=self.test_case.category.product)
+        self.test_case.add_component(self.good_component)
+        self.test_case.add_component(self.bad_component)
+
+    def verify_api_with_permission(self):
+        self.rpc_client.TestCase.remove_component(self.test_case.pk,
+                                                  self.bad_component.pk)
+        result = self.test_case.component.all()
+        self.assertEqual(len(result), 1)
+        self.assertEqual(result[0], self.good_component)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:639]: result.first() is better

File: tcms/rpc/tests/test_testcase.py [hunk 27]
```diff
@@ -522,16 +620,47 @@ def test_add_component_from_another_product_is_not_allowed(self):
...
+        self.good_component = ComponentFactory(product=self.test_case.category.product)
+        self.bad_component = ComponentFactory(product=self.test_case.category.product)
+        self.test_case.add_component(self.good_component)
+        self.test_case.add_component(self.bad_component)
+
+    def verify_api_with_permission(self):
+        self.rpc_client.TestCase.remove_component(self.test_case.pk,
+                                                  self.bad_component.pk)
+        result = self.test_case.component.all()
+        self.assertEqual(len(result), 1)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:638]: result.count() is better (and faster) b/c len() will evaluate the entire queryset and then count how many objects are there while .count() will do the counting directly in the database.

File: tcms/rpc/tests/test_testcase.py [hunk 28]
```diff
@@ -64,6 +136,11 @@ def _fixture_setup(self):
                 plan=[self.plan])
             self.cases.append(test_case)
 
+    def test_filter_query_none(self):
+        result = self.rpc_client.TestCase.filter()
+        self.assertIsNotNone(result)
+        self.assertNotEqual(len(result), 0)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:142]: assertGreater is better

File: tcms/rpc/tests/test_testcase.py [hunk 29]
```diff
@@ -569,14 +740,19 @@ def verify_api_with_permission(self):
         self.assertEqual(result[str(self.case_2.pk)], 15)
         self.assertEqual(result[str(self.case_3.pk)], 25)
 
+        # Test query None
+        result = self.rpc_client.TestCase.sortkeys()
+        self.assertIsNotNone(result)
+        self.assertNotEqual(len(result), 0)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:746]: assertGreater is better.

**Review by atodorov — CHANGES_REQUESTED** [review]
minor changes requested.

File: tcms/rpc/tests/test_testcase.py (hunk 27 above)
Comment (pandafy) [reply tcms/rpc/tests/test_testcase.py:638]: As far as I know, if we use count, then two queries to the database will be made - one for `count` and another for `result.first()`. Using `len` will evaluate the query and `result.first()` will be served from the cache.

Please let me know if I have the wrong understanding of the Queryset API I will make the requested changes. But, I think current implementation is quicker.

File: tcms/rpc/tests/test_testcase.py (hunk 25 above)
Comment (pandafy) [reply tcms/rpc/tests/test_testcase.py:523]: I got a little confused and changed the name of first return value. Both of them are number of deletions with respect to different things. :sweat_smile: 

I have fixed it. :smile:

File: tcms/rpc/tests/test_testcase.py (hunk 27 above)
Comment (atodorov) [reply tcms/rpc/tests/test_testcase.py:638]: Counting in DB is always quicker than counting in Python plus we like readability. Given `result` is of type queryset then use it as it was intended to be used.

File: tcms/rpc/tests/test_testcase.py [hunk 30]
```diff
@@ -16,29 +21,96 @@
...
+class TestValidateEmail(unittest.TestCase):
+    def test_non_list_email_collection(self):
+        with self.assertRaisesRegex(TypeError, 'cc_list should be a list object.'):
+            _validate_cc_list('example@example.com')
+
+    def test_invalid_email(self):
+        with self.assertRaises(ValidationError):
+            _validate_cc_list(['@example.com'])
+
+    def test_valid_email_list(self):  # pylint: disable=R0201
```
Comment (atodorov) [inline tcms/rpc/tests/test_testcase.py:33]: please use the human-readable message ID when disabling pylint errors. was that `no-self-use` ?

**Review by atodorov — CHANGES_REQUESTED** [review]
Almost there, 1 item remaining

**Review by atodorov — APPROVED** [review]
